beacon-chain/rpc/validator/proposer_utils.go

```diff
@@ -49,3 +49,47 @@ func (al proposerAtts) limitToMaxAttestations() proposerAtts {
 	}
 	return al
 }
+
+// dedup removes duplicate attestations (ones with the same bits set on).
+// Important: not only exact duplicates are removed, but proper subsets are removed too
+// (their known bits are redundant and are already contained in their supersets).
+func (al proposerAtts) dedup() proposerAtts {
+	if len(al) < 2 {
+		return al
+	}
+	attsByDataRoot := make(map[[32]byte][]*ethpb.Attestation, len(al))
+	for _, att := range al {
+		attDataRoot, err := att.Data.HashTreeRoot()
+		if err != nil {
+			continue
+		}
+		attsByDataRoot[attDataRoot] = append(attsByDataRoot[attDataRoot], att)
+	}
+
+	uniqAtts := make([]*ethpb.Attestation, 0, len(al))
+	for _, atts := range attsByDataRoot {
+		for i := 0; i < len(atts); i++ {
+			a := atts[i]
+			for j := i + 1; j < len(atts); j++ {
+				b := atts[j]
+				if a.AggregationBits.Contains(b.AggregationBits) {
+					// a contains b, b is redundant.
+					atts[j] = atts[len(atts)-1]
+					atts[len(atts)-1] = nil
+					atts = atts[:len(atts)-1]
+					j--
```

Comment:
Are you missing a `break` after this line?

Reply:
No, definitely not.

We fixate on `atts[i]` and can remove as many `atts[j]` w/o breaking (I am swapping the current element for the last and decrease the len -- common idiom). 

The break in the second `else if` is because when we need to remove `atts[i]` (and we fixate on it), we swap the last and the current element, and need to break, and check that last element once again.

Reply:
Makes sense!